File: src/ocaml-output/FStar_Parser_Lexhelp.ml

open Prims
open FStar_Pervasives

let intern_string : Prims.string  ->  Prims.string = (

let strings = (FStar_Util.smap_create (Prims.parse_int "100"))
in (fun s -> (

let uu____6 = (FStar_Util.smap_try_find strings s)
in (match (uu____6) with
| FStar_Pervasives_Native.Some (res) -> begin
res
end
| FStar_Pervasives_Native.None -> begin
((FStar_Util.smap_add strings s s);
s;
)
end))))


let default_string_finish = (fun endm b s -> (

let uu____29 = (FStar_Bytes.unicode_bytes_as_string s)
in FStar_Parser_Parse.STRING (uu____29)))


let call_string_finish = (fun fin buf endm b -> (let _0_26 = (FStar_Bytes.close buf)
in (fin endm b _0_26)))


let add_string : FStar_Bytes.bytebuf  ->  Prims.string  ->  Prims.unit = (fun buf x -> (

let uu____77 = (FStar_Bytes.string_as_unicode_bytes x)
in (FStar_Bytes.emit_bytes buf uu____77)))


let add_int_char : FStar_Bytes.bytebuf  ->  Prims.int  ->  Prims.unit = (fun buf c -> ((FStar_Bytes.emit_int_as_byte buf (c mod (Prims.parse_int "256")));
(FStar_Bytes.emit_int_as_byte buf (c / (Prims.parse_int "256")));
))


let add_unichar : FStar_Bytes.bytebuf  ->  Prims.int  ->  Prims.unit = (fun buf c -> (add_int_char buf c))


let add_byte_char : FStar_Bytes.bytebuf  ->  FStar_BaseTypes.char  ->  Prims.unit = (fun buf c -> (add_int_char buf ((FStar_Util.int_of_char c) mod (Prims.parse_int "256"))))


let stringbuf_as_bytes : FStar_Bytes.bytebuf  ->  FStar_Bytes.bytes = (fun buf -> (

let bytes = (FStar_Bytes.close buf)
in (

let uu____101 = (

let uu____102 = (FStar_Bytes.length bytes)
in (uu____102 / (Prims.parse_int "2")))
in (FStar_Bytes.make (fun i -> (FStar_Bytes.get bytes (FStar_Mul.op_Star i (Prims.parse_int "2")))) uu____101))))


let stringbuf_is_bytes : FStar_Bytes.bytebuf  ->  Prims.bool = (fun buf -> (

let bytes = (FStar_Bytes.close buf)
in (

let ok = (FStar_Util.mk_ref true)
in ((

let uu____115 = (

let uu____116 = (

let uu____117 = (FStar_Bytes.length bytes)
in (uu____117 / (Prims.parse_int "2")))
in (uu____116 - (Prims.parse_int "1")))
in (FStar_Util.for_range (Prims.parse_int "0") uu____115 (fun i -> (

let uu____122 = (

let uu____123 = (FStar_Bytes.get bytes ((FStar_Mul.op_Star i (Prims.parse_int "2")) + (Prims.parse_int "1")))
in (uu____123 <> (Prims.parse_int "0")))
in (match (uu____122) with
| true -> begin
(FStar_ST.write ok false)
end
| uu____126 -> begin
()
end)))));
(FStar_ST.read ok);
))))


let trigraph : FStar_BaseTypes.char  ->  FStar_BaseTypes.char  ->  FStar_BaseTypes.char  ->  FStar_BaseTypes.char = (fun c1 c2 c3 -> (

let digit = (fun c -> ((FStar_Util.int_of_char c) - (FStar_Util.int_of_char 48 (*0*))))
in (FStar_Util.char_of_int (((FStar_Mul.op_Star (digit c1) (Prims.parse_int "100")) + (FStar_Mul.op_Star (digit c2) (Prims.parse_int "10"))) + (digit c3)))))


let digit : FStar_BaseTypes.char  ->  Prims.int = (fun d -> (

let dd = (FStar_Util.int_of_char d)
in (match (((dd >= (FStar_Util.int_of_char 48 (*0*))) && (dd <= (FStar_Util.int_of_char 57 (*9*))))) with
| true -> begin
((FStar_Util.int_of_char d) - (FStar_Util.int_of_char 48 (*0*)))
end
| uu____146 -> begin
(failwith "digit")
end)))


let hexdigit : FStar_BaseTypes.char  ->  Prims.int = (fun d -> (

let dd = (FStar_Util.int_of_char d)
in (match (((dd >= (FStar_Util.int_of_char 48 (*0*))) && (dd <= (FStar_Util.int_of_char 57 (*9*))))) with
| true -> begin
(digit d)
end
| uu____151 -> begin
(match (((dd >= (FStar_Util.int_of_char 97 (*a*))) && (dd <= (FStar_Util.int_of_char 102 (*f*))))) with
| true -> begin
((dd - (FStar_Util.int_of_char 97 (*a*))) + (Prims.parse_int "10"))
end
| uu____152 -> begin
(match (((dd >= (FStar_Util.int_of_char 65 (*A*))) && (dd <= (FStar_Util.int_of_char 70 (*F*))))) with
| true -> begin
((dd - (FStar_Util.int_of_char 65 (*A*))) + (Prims.parse_int "10"))
end
| uu____153 -> begin
(failwith "hexdigit")
end)
end)
end)))


let unicodegraph_short : Prims.string  ->  FStar_BaseTypes.uint16 = (fun s -> (match (((FStar_String.length s) <> (Prims.parse_int "4"))) with
| true -> begin
(failwith "unicodegraph")
end
| uu____159 -> begin
(

let uu____160 = (

let uu____161 = (

let uu____162 = (

let uu____163 = (

let uu____164 = (

let uu____165 = (FStar_Util.char_at s (Prims.parse_int "0"))
in (hexdigit uu____165))
in (FStar_Mul.op_Star uu____164 (Prims.parse_int "4096")))
in (

let uu____166 = (

let uu____167 = (

let uu____168 = (FStar_Util.char_at s (Prims.parse_int "1"))
in (hexdigit uu____168))
in (FStar_Mul.op_Star uu____167 (Prims.parse_int "256")))
in (uu____163 + uu____166)))
in (

let uu____169 = (

let uu____170 = (

let uu____171 = (FStar_Util.char_at s (Prims.parse_int "2"))
in (hexdigit uu____171))
in (FStar_Mul.op_Star uu____170 (Prims.parse_int "16")))
in (uu____162 + uu____169)))
in (

let uu____172 = (

let uu____173 = (FStar_Util.char_at s (Prims.parse_int "3"))
in (hexdigit uu____173))
in (uu____161 + uu____172)))
in (FStar_Util.uint16_of_int uu____160))
end))


let hexgraph_short : Prims.string  ->  FStar_BaseTypes.uint16 = (fun s -> (match (((FStar_String.length s) <> (Prims.parse_int "2"))) with
| true -> begin
(failwith "hexgraph")
end
| uu____179 -> begin
(

let uu____180 = (

let uu____181 = (

let uu____182 = (

let uu____183 = (FStar_Util.char_at s (Prims.parse_int "0"))
in (hexdigit uu____183))
in (FStar_Mul.op_Star uu____182 (Prims.parse_int "16")))
in (

let uu____184 = (

let uu____185 = (FStar_Util.char_at s (Prims.parse_int "1"))
in (hexdigit uu____185))
in (uu____181 + uu____184)))
in (FStar_Util.uint16_of_int uu____180))
end))


let unicodegraph_long : Prims.string  ->  (FStar_BaseTypes.uint16 FStar_Pervasives_Native.option * FStar_BaseTypes.uint16) = (fun s -> (match (((FStar_String.length s) <> (Prims.parse_int "8"))) with
| true -> begin
(failwith "unicodegraph_long")
end
| uu____200 -> begin
(

let high = (

let uu____202 = (

let uu____203 = (

let uu____204 = (

let uu____205 = (

let uu____206 = (FStar_Util.char_at s (Prims.parse_int "0"))
in (hexdigit uu____206))
in (FStar_Mul.op_Star uu____205 (Prims.parse_int "4096")))
in (

let uu____207 = (

let uu____208 = (

let uu____209 = (FStar_Util.char_at s (Prims.parse_int "1"))
in (hexdigit uu____209))
in (FStar_Mul.op_Star uu____208 (Prims.parse_int "256")))
in (uu____204 + uu____207)))
in (

let uu____210 = (

let uu____211 = (

let uu____212 = (FStar_Util.char_at s (Prims.parse_int "2"))
in (hexdigit uu____212))
in (FStar_Mul.op_Star uu____211 (Prims.parse_int "16")))
in (uu____203 + uu____210)))
in (

let uu____213 = (

let uu____214 = (FStar_Util.char_at s (Prims.parse_int "3"))
in (hexdigit uu____214))
in (uu____202 + uu____213)))
in (

let low = (

let uu____216 = (

let uu____217 = (

let uu____218 = (

let uu____219 = (

let uu____220 = (FStar_Util.char_at s (Prims.parse_int "4"))
in (hexdigit uu____220))
in (FStar_Mul.op_Star uu____219 (Prims.parse_int "4096")))
in (

let uu____221 = (

let uu____222 = (

let uu____223 = (FStar_Util.char_at s (Prims.parse_int "5"))
in (hexdigit uu____223))
in (FStar_Mul.op_Star uu____222 (Prims.parse_int "256")))
in (uu____218 + uu____221)))
in (

let uu____224 = (

let uu____225 = (

let uu____226 = (FStar_Util.char_at s (Prims.parse_int "6"))
in (hexdigit uu____226))
in (FStar_Mul.op_Star uu____225 (Prims.parse_int "16")))
in (uu____217 + uu____224)))
in (

let uu____227 = (

let uu____228 = (FStar_Util.char_at s (Prims.parse_int "7"))
in (hexdigit uu____228))
in (uu____216 + uu____227)))
in (match ((high = (Prims.parse_int "0"))) with
| true -> begin
((FStar_Pervasives_Native.None), ((FStar_Util.uint16_of_int low)))
end
| uu____233 -> begin
((FStar_Pervasives_Native.Some ((FStar_Util.uint16_of_int ((Prims.parse_int "0xD800") + ((((FStar_Mul.op_Star high (Prims.parse_int "0x10000")) + low) - (Prims.parse_int "0x10000")) / (Prims.parse_int "0x400")))))), ((FStar_Util.uint16_of_int ((Prims.parse_int "0xDF30") + ((((FStar_Mul.op_Star high (Prims.parse_int "0x10000")) + low) - (Prims.parse_int "0x10000")) mod (Prims.parse_int "0x400"))))))
end)))
end))


let escape : FStar_Char.char  ->  FStar_Char.char = (fun c -> (match (c) with
| 92 (*\*) -> begin
92 (*\*)
end
| 39 (*'*) -> begin
39 (*'*)
end
| 110 (*n*) -> begin
10
end
| 116 (*t*) -> begin
9
end
| 98 (*b*) -> begin
8
end
| 114 (*r*) -> begin
13
end
| c1 -> begin
c1
end))

type compatibilityMode =
| ALWAYS
| FSHARP


let uu___is_ALWAYS : compatibilityMode  ->  Prims.bool = (fun projectee -> (match (projectee) with
| ALWAYS -> begin
true
end
| uu____242 -> begin
false
end))


let uu___is_FSHARP : compatibilityMode  ->  Prims.bool = (fun projectee -> (match (projectee) with
| FSHARP -> begin
true
end
| uu____246 -> begin
false
end))


let keywords : (compatibilityMode * Prims.string * FStar_Parser_Parse.token) Prims.list = (((ALWAYS), ("abstract"), (FStar_Parser_Parse.ABSTRACT)))::(((ALWAYS), ("attributes"), (FStar_Parser_Parse.ATTRIBUTES)))::(((ALWAYS), ("noeq"), (FStar_Parser_Parse.NOEQUALITY)))::(((ALWAYS), ("unopteq"), (FStar_Parser_Parse.UNOPTEQUALITY)))::(((ALWAYS), ("and"), (FStar_Parser_Parse.AND)))::(((ALWAYS), ("assert"), (FStar_Parser_Parse.ASSERT)))::(((ALWAYS), ("assume"), (FStar_Parser_Parse.ASSUME)))::(((ALWAYS), ("begin"), (FStar_Parser_Parse.BEGIN)))::(((ALWAYS), ("by"), (FStar_Parser_Parse.BY)))::(((FSHARP), ("default"), (FStar_Parser_Parse.DEFAULT)))::(((ALWAYS), ("effect"), (FStar_Parser_Parse.EFFECT)))::(((ALWAYS), ("else"), (FStar_Parser_Parse.ELSE)))::(((ALWAYS), ("end"), (FStar_Parser_Parse.END)))::(((ALWAYS), ("ensures"), (FStar_Parser_Parse.ENSURES)))::(((ALWAYS), ("exception"), (FStar_Parser_Parse.EXCEPTION)))::(((ALWAYS), ("exists"), (FStar_Parser_Parse.EXISTS)))::(((ALWAYS), ("false"), (FStar_Parser_Parse.FALSE)))::(((ALWAYS), ("False"), (FStar_Parser_Parse.L_FALSE)))::(((ALWAYS), ("forall"), (FStar_Parser_Parse.FORALL)))::(((ALWAYS), ("fun"), (FStar_Parser_Parse.FUN)))::(((ALWAYS), ("function"), (FStar_Parser_Parse.FUNCTION)))::(((ALWAYS), ("if"), (FStar_Parser_Parse.IF)))::(((ALWAYS), ("in"), (FStar_Parser_Parse.IN)))::(((ALWAYS), ("include"), (FStar_Parser_Parse.INCLUDE)))::(((ALWAYS), ("inline"), (FStar_Parser_Parse.INLINE)))::(((ALWAYS), ("inline_for_extraction"), (FStar_Parser_Parse.INLINE_FOR_EXTRACTION)))::(((ALWAYS), ("irreducible"), (FStar_Parser_Parse.IRREDUCIBLE)))::(((ALWAYS), ("let"), (FStar_Parser_Parse.LET (false))))::(((ALWAYS), ("logic"), (FStar_Parser_Parse.LOGIC)))::(((ALWAYS), ("match"), (FStar_Parser_Parse.MATCH)))::(((ALWAYS), ("module"), (FStar_Parser_Parse.MODULE)))::(((ALWAYS), ("mutable"), (FStar_Parser_Parse.MUTABLE)))::(((ALWAYS), ("new"), (FStar_Parser_Parse.NEW)))::(((ALWAYS), ("new_effect"), (FStar_Parser_Parse.NEW_EFFECT)))::(((ALWAYS), ("noextract"), (FStar_Parser_Parse.NOEXTRACT)))::(((ALWAYS), ("of"), (FStar_Parser_Parse.OF)))::(((ALWAYS), ("open"), (FStar_Parser_Parse.OPEN)))::(((ALWAYS), ("opaque"), (FStar_Parser_Parse.OPAQUE)))::(((ALWAYS), ("private"), (FStar_Parser_Parse.PRIVATE)))::(((ALWAYS), ("rec"), (FStar_Parser_Parse.REC)))::(((ALWAYS), ("reifiable"), (FStar_Parser_Parse.REIFIABLE)))::(((ALWAYS), ("reify"), (FStar_Parser_Parse.REIFY)))::(((ALWAYS), ("reflectable"), (FStar_Parser_Parse.REFLECTABLE)))::(((ALWAYS), ("requires"), (FStar_Parser_Parse.REQUIRES)))::(((ALWAYS), ("sub_effect"), (FStar_Parser_Parse.SUB_EFFECT)))::(((ALWAYS), ("then"), (FStar_Parser_Parse.THEN)))::(((ALWAYS), ("total"), (FStar_Parser_Parse.TOTAL)))::(((ALWAYS), ("true"), (FStar_Parser_Parse.TRUE)))::(((ALWAYS), ("True"), (FStar_Parser_Parse.L_TRUE)))::(((ALWAYS), ("try"), (FStar_Parser_Parse.TRY)))::(((ALWAYS), ("type"), (FStar_Parser_Parse.TYPE)))::(((ALWAYS), ("unfold"), (FStar_Parser_Parse.UNFOLD)))::(((ALWAYS), ("unfoldable"), (FStar_Parser_Parse.UNFOLDABLE)))::(((ALWAYS), ("val"), (FStar_Parser_Parse.VAL)))::(((ALWAYS), ("when"), (FStar_Parser_Parse.WHEN)))::(((ALWAYS), ("with"), (FStar_Parser_Parse.WITH)))::(((ALWAYS), ("_"), (FStar_Parser_Parse.UNDERSCORE)))::[]


let stringKeywords : Prims.string Prims.list = (FStar_List.map (fun uu____429 -> (match (uu____429) with
| (uu____433, w, uu____435) -> begin
w
end)) keywords)


let unreserve_words : Prims.string Prims.list = (FStar_List.choose (fun uu____440 -> (match (uu____440) with
| (mode, keyword, uu____447) -> begin
(match ((mode = FSHARP)) with
| true -> begin
FStar_Pervasives_Native.Some (keyword)
end
| uu____449 -> begin
FStar_Pervasives_Native.None
end)
end)) keywords)


let kwd_table : FStar_Parser_Parse.token FStar_Util.smap = (

let tab = (FStar_Util.smap_create (Prims.parse_int "1000"))
in ((FStar_List.iter (fun uu____457 -> (match (uu____457) with
| (mode, keyword, token) -> begin
(FStar_Util.smap_add tab keyword token)
end)) keywords);
tab;
))


let kwd : Prims.string  ->  FStar_Parser_Parse.token FStar_Pervasives_Native.option = (fun s -> (FStar_Util.smap_try_find kwd_table s))

type lexargs =
{getSourceDirectory : Prims.unit  ->  Prims.string; filename : Prims.string; contents : Prims.string}


let mkLexargs : ((Prims.unit  ->  Prims.string) * Prims.string * Prims.string)  ->  lexargs = (fun uu____508 -> (match (uu____508) with
| (srcdir, filename, contents) -> begin
{getSourceDirectory = srcdir; filename = filename; contents = contents}
end))


let kwd_or_id : lexargs  ->  FStar_Range.range  ->  Prims.string  ->  FStar_Parser_Parse.token = (fun args r s -> (

let uu____530 = (kwd s)
in (match (uu____530) with
| FStar_Pervasives_Native.Some (v1) -> begin
v1
end
| FStar_Pervasives_Native.None -> begin
(match (s) with
| "__SOURCE_DIRECTORY__" -> begin
(

let uu____533 = (args.getSourceDirectory ())
in FStar_Parser_Parse.STRING (uu____533))
end
| "__SOURCE_FILE__" -> begin
(

let uu____534 = (FStar_Range.file_of_range r)
in FStar_Parser_Parse.STRING (uu____534))
end
| "__LINE__" -> begin
(

let uu____535 = (

let uu____538 = (

let uu____539 = (

let uu____540 = (FStar_Range.start_of_range r)
in (FStar_Range.line_of_pos uu____540))
in (FStar_All.pipe_left FStar_Util.string_of_int uu____539))
in ((uu____538), (false)))
in FStar_Parser_Parse.INT (uu____535))
end
| uu____541 -> begin
(match ((FStar_Util.starts_with s FStar_Ident.reserved_prefix)) with
| true -> begin
(FStar_Pervasives.raise (FStar_Errors.Error ((((Prims.strcat FStar_Ident.reserved_prefix " is a reserved prefix for an identifier")), (r)))))
end
| uu____542 -> begin
(

let uu____543 = (intern_string s)
in FStar_Parser_Parse.IDENT (uu____543))
end)
end)
end)))

File: src/parser/FStar.Parser.Lexhelp.fs
(*
   Copyright 2008-2014 Nikhil Swamy and Microsoft Research

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*)
// Adapted from sources of F#
//----------------------------------------------------------------------------
//
// Copyright (c) 2002-2012 Microsoft Corporation.
//
// This source code is subject to terms and conditions of the Apache License, Version 2.0. A
// copy of the license can be found in the License.html file at the root of this distribution.
// By using this source code in any fashion, you are agreeing to be bound
// by the terms of the Apache License, Version 2.0.
//
// You must not remove this notice, or any other, from this software.
//----------------------------------------------------------------------------
// See LICENSE-fsharp.txt at the root of this distribution for terms and conditions
#light "off"

module FStar.Parser.Lexhelp
open FStar.ST
open FStar.Exn
open FStar.All
open FStar.Mul
open FStar
open FStar.Util
open FStar.Range
open FStar.Errors
open FStar.Parser
open FStar.Parser.Parse
open FStar.BaseTypes
open FStar.Compiler

let intern_string : string -> string =
  let strings = Util.smap_create 100 in
  fun s ->
    match Util.smap_try_find strings s with
      | Some res -> res
      | None -> Util.smap_add strings s s; s

let default_string_finish endm b s = STRING (Bytes.unicode_bytes_as_string s)

let call_string_finish fin buf endm b = fin endm b (Bytes.close buf)

let add_string buf x = Bytes.emit_bytes buf (Bytes.string_as_unicode_bytes x)

let add_int_char buf c =
  Bytes.emit_int_as_byte buf (c % 256);
  Bytes.emit_int_as_byte buf (c / 256)

let add_unichar buf c = add_int_char buf c
let add_byte_char buf (c:char) = add_int_char buf (Util.int_of_char c % 256)

(* When lexing bytearrays we don't expect to see any unicode stuff. *)
(* Likewise when lexing string constants we shouldn't see any trigraphs > 127 *)
(* So to turn the bytes collected in the string buffer back into a bytearray *)
(* we just take every second byte we stored.  Note all bytes > 127 should have been *)
(* stored using add_int_char *)
let stringbuf_as_bytes buf =
    let bytes = Bytes.close buf in
    Bytes.make (fun i -> Bytes.get bytes (i*2)) (Bytes.length bytes / 2)

(* Sanity check that high bytes are zeros. Further check each low byte <= 127 *)
let stringbuf_is_bytes buf =
    let bytes = Bytes.close buf in
    let ok = Util.mk_ref true in
    Util.for_range 0 (Bytes.length bytes/2-1) (fun i ->
      if Bytes.get bytes (i*2+1) <> 0
      then ok := false
      else ());
    !ok

let trigraph c1 c2 c3 =
    let digit (c:char) = Util.int_of_char c - Util.int_of_char '0' in
    char_of_int (digit c1 * 100 + digit c2 * 10 + digit c3)

let digit d =
    let dd = int_of_char d in
    if dd >= int_of_char '0' && dd <= int_of_char '9' then int_of_char d - int_of_char '0'
    else failwith "digit"

let hexdigit d =
    let dd = int_of_char d in
    if dd >= int_of_char '0' && dd <= int_of_char '9' then digit d
    else if dd >= int_of_char 'a' && dd <= int_of_char 'f' then dd - int_of_char 'a' + 10
    else if dd >= int_of_char 'A' && dd <= int_of_char 'F' then dd - int_of_char 'A' + 10
    else failwith "hexdigit"

let unicodegraph_short s =
    if String.length s <> 4
    then failwith "unicodegraph"
    else uint16_of_int (hexdigit (char_at s 0) * 4096 + hexdigit (char_at s 1) * 256 + hexdigit (char_at s 2) * 16 + hexdigit (char_at s 3))

let hexgraph_short s =
    if String.length s <> 2
    then failwith "hexgraph"
    else uint16_of_int (hexdigit (char_at s 0) * 16 + hexdigit (char_at s 1))

let unicodegraph_long s =
    if String.length s <> 8
    then failwith "unicodegraph_long"
    else
      let high = hexdigit (char_at s 0) * 4096 + hexdigit (char_at s 1) * 256 + hexdigit (char_at s 2) * 16 + hexdigit (char_at s 3) in
      let low = hexdigit (char_at s 4) * 4096 + hexdigit (char_at s 5) * 256 + hexdigit (char_at s 6) * 16 + hexdigit (char_at s 7) in
      if high = 0 then None, uint16_of_int low
      else
      // A surrogate pair - see http://www.unicode.org/unicode/uni2book/ch03.pdf, section 3.7
        Some (uint16_of_int (0xD800 + ((high * 0x10000 + low - 0x10000) / 0x400))),
        uint16_of_int (0xDF30 + ((high * 0x10000 + low - 0x10000) % 0x400))

let escape c =
    match c with
    | '\\' -> '\\'
    | '\'' -> '\''
    | 'n' -> '\n'
    | 't' -> '\t'
    | 'b' -> '\b'
    | 'r' -> '\r'
    | c -> c

(*------------------------------------------------------------------------
!* Keyword table
 *-----------------------------------------------------------------------*)

type compatibilityMode =
    | ALWAYS  (* keyword *)
    | FSHARP  (* keyword, but an identifier under --ml-compatibility mode *)

let keywords =
  [ ALWAYS, "abstract"   ,ABSTRACT;
    ALWAYS, "attributes" ,ATTRIBUTES;
    ALWAYS, "noeq"       ,NOEQUALITY;
    ALWAYS, "unopteq"    ,UNOPTEQUALITY;
    ALWAYS, "and"        ,AND;
    ALWAYS, "assert"     ,ASSERT;
    ALWAYS, "assume"     ,ASSUME;
    ALWAYS, "begin"      ,BEGIN;
    ALWAYS, "by"         ,BY;
    ALWAYS, "calc"       ,CALC;
    ALWAYS, "class"      ,CLASS;
    FSHARP, "default"    ,DEFAULT;
    ALWAYS, "effect"     ,EFFECT;
    ALWAYS, "else"       ,ELSE;
    ALWAYS, "end"        ,END;
    ALWAYS, "ensures"    ,ENSURES;
    ALWAYS, "exception"  ,EXCEPTION;
    ALWAYS, "exists"     ,EXISTS;
    ALWAYS, "false"      ,FALSE;
    ALWAYS, "friend"     ,FRIEND;
    ALWAYS, "forall"     ,FORALL;
    ALWAYS, "fun"        ,FUN;
    ALWAYS, "function"   ,FUNCTION;
    ALWAYS, "if"         ,IF;
    ALWAYS, "in"         ,IN;
    ALWAYS, "include"    ,INCLUDE;
    ALWAYS, "inline"     ,INLINE;
    ALWAYS, "inline_for_extraction"     ,INLINE_FOR_EXTRACTION;
    ALWAYS, "instance"   ,INSTANCE;
    ALWAYS, "irreducible",IRREDUCIBLE;
    ALWAYS, "let"        ,LET(false);
    ALWAYS, "logic"      ,LOGIC;
    ALWAYS, "match"      ,MATCH;
    ALWAYS, "module"     ,MODULE;
    ALWAYS, "new"        ,NEW;
    ALWAYS, "new_effect" ,NEW_EFFECT;
    ALWAYS, "layered_effect"            ,LAYERED_EFFECT;
    ALWAYS, "noextract",  NOEXTRACT;
    ALWAYS, "of"         ,OF;
    ALWAYS, "open"       ,OPEN;
    ALWAYS, "opaque"     ,OPAQUE;
    ALWAYS, "private"    ,PRIVATE;
    ALWAYS, "range_of"   ,RANGE_OF;
    ALWAYS, "rec"        ,REC;
    ALWAYS, "reifiable"  ,REIFIABLE;
    ALWAYS, "reify"      ,REIFY;
    ALWAYS, "reflectable",REFLECTABLE;
    ALWAYS, "requires"   ,REQUIRES;
    ALWAYS, "set_range_of",SET_RANGE_OF;
    ALWAYS, "sub_effect" ,SUB_EFFECT;
    ALWAYS, "synth"      ,SYNTH;
    ALWAYS, "then"       ,THEN;
    ALWAYS, "total"      ,TOTAL;
    ALWAYS, "true"       ,TRUE;
    ALWAYS, "try"        ,TRY;
    ALWAYS, "type"       ,TYPE;
    ALWAYS, "unfold"     ,UNFOLD;
    ALWAYS, "unfoldable" ,UNFOLDABLE;
    ALWAYS, "val"        ,VAL;
    ALWAYS, "when"       ,WHEN;
    ALWAYS, "with"       ,WITH;
    ALWAYS, "_"          ,UNDERSCORE;
  ]
let stringKeywords = List.map (fun (_, w, _) -> w) keywords

(*------------------------------------------------------------------------
!* Keywords
 *-----------------------------------------------------------------------*)

let unreserve_words =
    List.choose (fun (mode,keyword,_) -> if mode = FSHARP then Some keyword else None) keywords

let kwd_table =
    let tab = Util.smap_create 1000 in
    List.iter (fun (mode,keyword,token) -> Util.smap_add tab keyword token) keywords;
    tab
let kwd s = Util.smap_try_find kwd_table s

type lexargs = {
  getSourceDirectory: (unit -> string);
  filename:string;
  contents:string
}

let mkLexargs (srcdir,(filename:string),(contents:string)) = {
  getSourceDirectory = srcdir;
  filename = filename;
  contents = contents
}

let kwd_or_id args (r:Range.range) s =
  match kwd s with
    | Some v ->
      v
    | None ->
      match s with
        | "__SOURCE_DIRECTORY__" ->
          STRING (args.getSourceDirectory())
        | "__SOURCE_FILE__" ->
          STRING (Range.file_of_range r)
        | "__LINE__" ->
          INT (Util.string_of_int <| Range.line_of_pos (Range.start_of_range r), false)
        | _ ->
          if Util.starts_with s Ident.reserved_prefix
          then raise_error (Errors.Fatal_ReservedPrefix, (Ident.reserved_prefix  ^ " is a reserved prefix for an identifier")) r
          else IDENT (intern_string(s))
